pkg/scalers/azure_pipelines_scaler.go

```diff
@@ -243,47 +330,50 @@ func (s *azurePipelinesScaler) GetAzurePipelinesQueueLength(ctx context.Context)
 		return -1, err
 	}
 
-	var result map[string]interface{}
-	err = json.Unmarshal(body, &result)
+	var jrs JobRequests
+	err = json.Unmarshal(body, &jrs)
 	if err != nil {
+		s.logger.Error(err, "Cannot unmarshal ADO JobRequests API response")
 		return -1, err
 	}
 
-	var count int64
-	jobs, ok := result["value"].([]interface{})
-
-	if !ok {
-		return -1, fmt.Errorf("the Azure DevOps REST API result returned no value data despite successful code. url: %s", url)
-	}
-
 	// for each job check if it parent fulfilled, then demand fulfilled, then finally pool fulfilled
-	for _, value := range jobs {
-		v := value.(map[string]interface{})
-		if v["result"] == nil {
-			if s.metadata.parent == "" && s.metadata.demands == "" {
-				// no plan defined, just add a count
-				count++
+	var count int64
+	for _, job := range stripDeadJobs(jrs.Value) {
+		if s.metadata.parent == "" && s.metadata.demands == "" {
+			// no plan defined, just add a count
+			count++
+		} else {
+			if s.metadata.parent == "" {
+				// doesn't use parent, switch to demand
+				if getCanAgentDemandFulfilJob(job, s.metadata) {
+					count++
+				}
 			} else {
-				if s.metadata.parent == "" {
-					// doesn't use parent, switch to demand
-					if getCanAgentDemandFulfilJob(v, s.metadata) {
-						count++
-					}
-				} else {
-					// does use parent
-					if getCanAgentParentFulfilJob(v, s.metadata) {
-						count++
-					}
+				// does use parent
+				if getCanAgentParentFulfilJob(job, s.metadata) {
+					count++
 				}
 			}
 		}
 	}
+
 	return count, err
 }
 
+func stripDeadJobs(jobs []JobRequest) []JobRequest {
+	var filtered []JobRequest
+	for _, job := range jobs {
+		if job.Result == nil {
+			filtered = append(filtered, job)
+		}
+	}
+	return filtered
+}
```

Comment:
I believe the compiler breaks it down into 2 separate iterations; the numbers show that this is faster on a queue of circa 1k jobs (normal). I can give more proof if required :)

Reply:
more than introducing
```
if job.Result == nil {
	filtered = append(filtered, job)
}
``` 

In the loop where this is called directly ?(line 343) 
Interesting, I thought that iterating over all the elements and applying the condition inside the loop was faster than iterating once to filter the collection and then iterate again to process the filtered collection.

If you are sure, we can merge this as it is, honestly, I don't think that this will make a huge difference considering that the filtered collection will have only a few elements, it's not like iterating twice  1kk elements

Reply:
I mean, this isn't a blocker for me, it was just a question "for the interest" 😄

Reply:
I'll be honest, it was a surprise for me too. But the results show the strip works first. I devolved it to a separate function in case we need to add additional filters at a later date, but that may of course negatively impact performance. Going to resolve the conversation. :)